### ISSUE TITLE
Some browsers take a really long time to process large tables.

### ISSUE THREAD

**Issue by schutze** [hhurz/tableExport.jquery.plugin#195]
Similar to #132, I tested a 3 column table in IE 11.0.9600, Firefox 55.0.3, and Chrome 61 (results in seconds):

75 rows to CSV : IE 11 = 5, Firefox = 9, Chrome = 2
75 rows to XLS : IE 11 = 5, Firefox = 9, Chrome = 2
275 rows to CSV : IE 11 = 21, Firefox = 64, Chrome = 7
275 rows to XLS : IE 11 = 22, Firefox = 63, Chrome = 7
275 rows to PDF AutoTable: IE 11 = 57, Firefox = 64, Chrome = 7

Except for IE and AutoTable, the type of output does not seem to matter very much. Is there maybe some inefficient code that prepares data for multiple output types?

**Comment by prog20901** [comment]
To solve this problem, is it possible to add pagination? else one more option to include all these export options bundles to datatable including their existing ones. By default they have 5 options - copy to csv, clipboard, pdf, excel, print option. example see the front page of it https://datatables.net/extensions/buttons/examples/initialisation/export it would be awesome if we do that.....

The datatable works great even if the size is 10,000 where pagination is set

https://datatables.net/extensions/scroller/examples/initialisation/large_js_source.html

Integrating all the export features to datatable would be really awesome...

You can also get in touch with forum https://datatables.net/forums/ where huge people will respond to improve it if required. Other examples would be helpful

1. Export to Json example - https://datatables.net//forums/discussion/comment/92685/#Comment_92685
2. Responsive by single attribute - https://datatables.net/extensions/responsive/

**Comment by hhurz** [comment]
@schutze: can you provide your test code for further investigation?

@jkathir: Short answer is: no. This plugin requires that all table data is available on client side and filled into a html table. For more information see issue #187 and #182

**Comment by prog20901** [comment]
Datatable is also making all the data on the client side as well. They will work for static html files as well. The data doesn't needs to be dynamic. Since it has few good features in default - pagination, responsive, copy to clipboard, csv, print, pdf, excel, search i suggested for you to re-consider. It is up to you to decide whether to incorporate all these features in datatable or incorporate it to make efficient like datatable.

**Comment by schutze** [comment]
The slowness seems to be related to a highly stylized table.  If the CSS is not applied then it only takes the plugin about 3 seconds to produce the output.

I created a fiddle to demonstrate; https://jsfiddle.net/0Lgosz51/6/

Maybe there are code efficiencies that can be made to improve performance, at least for those output types, like CSV, that do not support output styling?

**Comment by hhurz** [comment]
@schutze  Thanks a lot for providing the sample code and for figuring out what slows things down. I found two jquery functions that have a poor performance if there is extensive usage of css styling. One function is jQuery css() and the other is jQuery is(). Both functions are used in this plugin for checking visibilty and for retrieving css information of rows and cells. Actually I started to find a solution for this issue.

**Comment by ggywebtech** [comment]
Hi Guys, I have a similar problem when I'm using bootstrap-table, only mine is not connected with the styling but the filter select option for each column. To get around this I have modified the bootstrap-table-export.js to remove the filter select row with $( ".fht-cell" ).detach(); before the export. Then of cause put it back after. If it helps here my fiddle is exporting 3600 rows.
[http://jsfiddle.net/apseb49f/9/](url)

**Comment by hhurz** [comment]
@schutze FYI: I just committed a fix for this issue. Export should work faster now...

**Comment by schutze** [comment]
@hhurz thanks for the update!  I applied the new tableExport.js and updated the fiddle.  Be sure to force a browser refresh so the script with the updated plugin is used.

I'm not sure if this plugin version was supposed to speed up every output format but (unformatted) CSV is much quicker and (formatted) XLS and PDF is mostly slower.

https://jsfiddle.net/0Lgosz51/6/

**Here are the previous results:**
275 rows to CSV : IE = 21, Firefox = 64, Chrome = 7
275 rows to Formatted XLS : IE 11 = 22, Firefox = 63, Chrome = 7
275 rows to PDF AutoTable: IE 11 = 57, Firefox = 64, Chrome = 7

**Here are the new results:**
275 rows to CSV : IE = 7, Firefox = 2, Chrome = 2
275 rows to Formatted XLS : IE 11 = 29, Firefox = 66, Chrome = 8
275 rows to PDF AutoTable: IE 11 = 27, Firefox = 66, Chrome = 16

I used IE 11.0.9600, Firefox 56.0, and Chrome 61

**Comment by schutze** [comment]
Since this was closed, I suppose a performance improvement for formatted output of highly stylized tables will not be considered?  The plugin is not very useful for this combination.  It is so slow to export tables with greater than 300 rows, that users may think nothing is happening.

**Comment by hhurz** [comment]
There is no way (at least for me) to optimize the plugins code for faster export of such "highly stylized" (I'd say "bad styled") tables. The last update only improves the export of those tables to formats which do not deal with css stylings. Both formats, Excel optionally and PDF always, access the css properties of the table, which extremely slows down the export process. So yes, if you are not able to optimize the original table's css styling but you want to export to Excel or to PDF, you shouldn't use this plugin.

**Comment by RNONGFU001** [comment]
@hhurz I also experienced the same problem. my table has over 2500 rows, and it really takes long time. And after reading all of above messages, I wonder if you have to check the visibility of the cell? I mean if you do not check the cell visibility, then I guess we will get all the cells of the table, right? but if it could improve the output performance, who cares?

**Comment by hhurz** [comment]
@RSHEN001 Does your table contain hidden elements (tr, th, td) ? If so, this could be one reason why the export process slows down. As mentioned above, extensive css styling of your table could be an other one.

**Comment by RNONGFU001** [comment]
@hhurz Yes, I hide some columns. But my question is if I marked the code doing the visibility check, does your code still work, and is it possible to speed up the process by doing so

**Comment by hhurz** [comment]
@RSHEN001  In release 1.9.8 you will find a new option 'exportHiddenCells'. Set this option to true and try to export your table.
